File: algorithms/0098. 验证二叉搜索树/0098.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @return {boolean}
 */
const isValidBST = function(root) {
    const helper = function(root, min, max) {
        if (!root) return true
        if ((min !== null && root.val <= min) || (max !== null && root.val >= max)) return false
            return helper(root.left, min, root.val) && helper(root.right, root.val, max)
    }
    return helper(root, null, null)
}

File: algorithms/0098. 验证二叉搜索树/0098.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None


class Solution:
    def isValidBST(self, root: TreeNode) -> bool:
        return self.helper(root, None, None)

    def helper(self, root: TreeNode, minVal, maxVal) -> bool:
        if not root:
            return True
        if (minVal is not None and root.val <= minVal):
            return False
        if (maxVal is not None and root.val >= maxVal):
            return False
        return self.helper(root.left, minVal, root.val) \
            and self.helper(root.right, root.val, maxVal)
